File: leetcode_c_plus_plus/141.cpp
using namespace std;
#include <vector>
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
 };
class Solution {
public:
    bool hasCycle(ListNode *head) {
        
        if(head == NULL || head->next == NULL){
            return false;
        }
        ListNode * slow = head ;
        ListNode * fast = head->next;
        while(fast != NULL){
            slow = slow->next;
            if(fast->next == NULL){
                return false;
            }
            fast = fast->next->next;
            if(slow == fast){
                return true;
            }
        }
        return false;
        
    }
};

File: leetcode_python/141.py
# 141. 环形链表
# https://leetcode-cn.com/problems/linked-list-cycle/
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        # 注意这里面的判断的
        if not head or not head.next:
            return False
        slow = head
        fast = head.next

        while slow != fast:
            # 注意这里面的结束条件
            if not fast or not fast.next:
                return False
            slow = slow.next
            fast = fast.next.next
        return True